hermes/bin/src/runtime_extensions/hermes/http_gateway/routing.rs

```diff
@@ -0,0 +1,168 @@
+    let unique_request_id = EventUID(rusty_ulid::generate_ulid_string());
+
+    connection_manager
+        .connection_context
+        .try_lock()
+        .map_err(|_| anyhow::anyhow!("Unable to obtain mutex lock"))?
+        .insert(
+            unique_request_id.clone(),
+            (ClientIPAddr(ip), Processed(false), LiveConnection(true)),
+        );
+
+    info!("connection manager {:?}", connection_manager);
+
+    let (_app_name, resolved_host) = host_resolver(req.headers())?;
+
+    let response = if config
+        .valid_hosts
+        .iter()
+        .any(|host| host.0 == resolved_host.0.as_str())
+    {
+        route_to_hermes(req).await?
+    } else {
+        return Ok(error_response("hostname not valid".to_owned()));
+    };
+
+    connection_manager
+        .connection_context
+        .try_lock()
+        .map_err(|_| anyhow::anyhow!("Unable to obtain mutex lock"))?
+        .insert(
+            unique_request_id,
+            (ClientIPAddr(ip), Processed(true), LiveConnection(false)),
+        );
+
+    info!("connection manager {:?}\n", connection_manager);
+
+    Ok(response)
+}
+
+/// Route single request to hermes backend
+async fn route_to_hermes(req: Request<Body>) -> anyhow::Result<Response<Body>> {
+    let (lambda_send, lambda_recv_answer): (Sender<HTTPEventMsg>, Receiver<HTTPEventMsg>) =
+        channel();
+
+    let uri = req.uri().to_owned();
+    let method = req.method().to_owned().to_string();
+    let path = req.uri().path().to_string();
+
+    let mut header_map: HashMap<String, Vec<String>> = HashMap::new();
+
+    for (header_name, header_val) in req.headers() {
+        header_map
+            .entry(header_name.to_string())
+            .or_default()
+            .push(header_val.to_str()?.to_string());
+    }
+
+    match uri.path() {
+        "/api" => {
+            compose_http_event(
+                method,
+                header_map.into_iter().collect(),
+                req.collect().await?.to_bytes(), // body
+                path,
+                lambda_send,
+                lambda_recv_answer,
+            )
+        },
+        _ => todo!(),
+    }
+}
+
+/// Compose http event and send to global queue, await queue response and relay back to
+/// waiting receiver channel for HTTP response
+fn compose_http_event(
+    method: String, headers: HeadersKV, body: Bytes, path: String, sender: Sender<HTTPEventMsg>,
+    receiver: Receiver<HTTPEventMsg>,
+) -> anyhow::Result<Response<Body>> {
+    let on_http_event = HTTPEvent {
+        headers,
+        method,
+        path,
+        body,
+        sender,
+    };
+    crate::event::queue::send(HermesEvent::new(
+        on_http_event,
+        TargetApp::All,
+        TargetModule::All,
+    ))?;
+
+    match &receiver.recv()? {
+        HTTPEventMsg::HttpEventResponseSome(resp) => {
+            Ok(Response::new(serde_json::to_string(&resp)?.into()))
+        },
+        HTTPEventMsg::HttpEventResponseNone() => Ok(not_found()),
```

Comment:
@stevenj in the case where the event responds with nothing (none). What should we return to the client? Should we classify nothing returned as a 404?

Reply:
We should not return anyting, UNTIL all wasm modules have finished.  At which time if none of them supplied a response we return 404.